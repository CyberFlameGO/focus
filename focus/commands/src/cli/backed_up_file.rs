use anyhow::{Context, Result};
use std::{
    cell::Cell,
    path::{Path, PathBuf},
};

#[derive(Debug)]
pub struct BackedUpFile {
    original_path: PathBuf,
    backup_path: PathBuf,
    restore: Cell<bool>,
}

impl BackedUpFile {
    pub fn new(path: &Path) -> Result<Self> {
        let mut backup_path = path.to_owned();
        let mut name = backup_path
            .file_name()
            .expect("Backup file with no file name")
            .to_owned();
        name.push(".backup");
        backup_path.set_file_name(name);

        std::fs::copy(&path, &backup_path).with_context(|| {
            format!(
                "Copying {} to the backup file {}",
                &path.display(),
                &backup_path.display()
            )
        })?;

        log::debug!(
            "Backed up {} -> {}",
            &path.display(),
            &backup_path.display()
        );

        Ok(Self {
            original_path: path.to_owned(),
            backup_path: backup_path.to_owned(),
            restore: Cell::new(true),
        })
    }

    pub fn set_restore(&self, new_value: bool) {
        self.restore.set(new_value);
    }
}

impl Drop for BackedUpFile {
    fn drop(&mut self) {
        if self.restore.get() {
            log::info!(
                "Restoring backup {} -> {}",
                self.backup_path.display(),
                self.original_path.display()
            );
            std::fs::rename(&self.backup_path, &self.original_path)
                .expect("failed to restore backup file");
        } else {
            log::debug!("Removing backup {}", self.backup_path.display());
            std::fs::remove_file(&self.backup_path).expect("failed to delete backup file");
        }
    }
}
