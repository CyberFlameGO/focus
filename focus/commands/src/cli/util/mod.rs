pub mod backed_up_file;
pub mod lock_file;
pub mod periodic;
pub mod sandbox;
pub mod sandbox_command;
pub mod temporary_working_directory;
