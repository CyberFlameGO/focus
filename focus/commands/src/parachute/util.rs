use std::path::{Path, PathBuf};

pub struct TemporaryWorkingDirectory {
    original_directory: PathBuf,
}

impl TemporaryWorkingDirectory {
    pub(crate) fn new(directory_to_switch_to: &Path) -> anyhow::Result<Self> {
        use std::env;
        use anyhow::Context;
        let current_dir = env::current_dir().context("getting the current directory failed")?;
        env::set_current_dir(directory_to_switch_to).context("switching to the new directory failed")?;
        Ok(Self {
            original_directory: current_dir,
        })
    }
}

impl Drop for TemporaryWorkingDirectory {
    fn drop(&mut self) {
        std::env::set_current_dir(&self.original_directory).expect("switching back to the original directory failed");
    }
}
