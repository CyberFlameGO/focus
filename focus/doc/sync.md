# Focus State Sync / Profile Application


## Ideas

- Complain in Bazel / calculate the stuff there

- Upstream changes that complain when the underyling profile changes and has not been reapplied

- Merge hook
	- How does IntelliJ do merges?
- Clippy + repo daemon watching files
	- Clippy reminds tells you to sync your project using desktop notifications or somesuch.

### Underpinnings
- `git diff --name-only <last_application_point>..`

