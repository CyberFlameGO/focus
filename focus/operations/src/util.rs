use anyhow::{bail, Result};

use tracing::error;

pub fn perform<F, J>(description: &str, f: F) -> Result<J>
where
    F: FnOnce() -> Result<J>,
{
    let result = f();
    if let Err(e) = result {
        error!("Failed {}: {}", description.to_ascii_lowercase(), e);
        bail!(e);
    }

    result
}
