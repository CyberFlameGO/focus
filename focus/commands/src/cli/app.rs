use std::sync::Arc;

use crate::{ui::UserInterface, util::sandbox::Sandbox};
use anyhow::{Context, Result};

#[derive(Clone)]
pub struct App {
    ui: Arc<UserInterface>,
    sandbox: Arc<Sandbox>,
}

impl App {
    pub fn new(preserve_sandbox_contents: bool, interactive: bool) -> Result<Self> {
        let ui = Arc::from(UserInterface::new(interactive).context("Failed to start UI")?);
        let sandbox =
            Arc::from(Sandbox::new(preserve_sandbox_contents).context("Failed to create sandbox")?);

        Ok(Self { ui, sandbox })
    }

    /// Get a reference to the app's ui.
    pub fn ui(&self) -> Arc<UserInterface> {
        self.ui.clone()
    }

    /// Get a reference to the app's sandbox.
    pub fn sandbox(&self) -> Arc<Sandbox> {
        self.sandbox.clone()
    }
}
