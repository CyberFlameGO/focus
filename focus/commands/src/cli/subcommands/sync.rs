use crate::app::App;
use crate::coordinate::CoordinateSet;
use crate::model::Layer;
use crate::model::LayerSets;
use crate::sandbox_command::SandboxCommand;
use std::convert::TryFrom;
use std::fs::File;

use std::path::Path;
use std::path::PathBuf;
use std::process::Stdio;
use std::sync::Arc;

use anyhow::{bail, Context, Result};

use crate::{
    sandbox_command::SandboxCommandOutput, working_tree_synchronizer::WorkingTreeSynchronizer,
};

pub fn perform<F, J>(description: &str, f: F) -> Result<J>
where
    F: FnOnce() -> Result<J>,
{
    let result = f();
    if let Err(e) = result {
        log::error!("Failed {}: {}", description.to_ascii_lowercase(), e);
        bail!(e);
    }

    result
}

pub fn run(app: Arc<App>, sparse_repo: &Path) -> Result<()> {
    use crate::git_helper;
    use crate::sparse_repos;

    let ui = app.ui();
    let sparse_repo = git_helper::find_top_level(app.clone(), &sparse_repo)
        .context("canonicalizing sparse repo path")?;
    let dense_repo = git_helper::run_git_command_consuming_stdout(
        "Reading dense repo URL".to_owned(),
        &sparse_repo,
        vec!["remote", "get-url", "dense"],
        app.clone(),
    )
    .context("Failed reading the dense repo URL")
    .map(|path| PathBuf::from(path))?;
    let dense_repo = git_helper::find_top_level(app.clone(), &dense_repo)
        .context("Failed finding dense repo top level")?;

    let _ = ui.status(format!(
        "Syncing {}",
        &sparse_repo.display(),
    ));

    let sparse_sync = WorkingTreeSynchronizer::new(&sparse_repo, app.clone())?;
    let dense_sync = WorkingTreeSynchronizer::new(&dense_repo, app.clone())?;

    let sparse_profile_path = sparse_repo
        .join(".git")
        .join("info")
        .join("sparse-checkout");

    let (sparse_profile_output_file, sparse_profile_output_path) =
        app.sandbox().create_file(Some("sparse-profile"), None)?;
    drop(sparse_profile_output_file);

    let sparse_checkout_backup_path = {
        let mut path = sparse_profile_path.clone();
        path.set_extension("backup");
        path
    };

    if let Ok(clean) = perform("Checking that dense repo is in a clean state", || {
        dense_sync.is_working_tree_clean()
    }) {
        if !clean {
            eprintln!("The working tree in the dense repo must be in a clean state. Commit or stash changes and try to run the sync again.");
            bail!("Dense repo working tree is not in a clean state");
        }
    } else {
        bail!("Could not determine whether the dense repo is in a clean state");
    }

    if let Ok(clean) = perform("Checking that sparse repo is in a clean state", || {
        sparse_sync.is_working_tree_clean()
    }) {
        if !clean {
            eprintln!("The working tree in the sparse repo must be in a clean state. Commit or stash changes and try to run the sync again.");
            bail!("Sparse repo working tree is not in a clean state");
        }
    } else {
        bail!("Could not determine whether the sparse repo is in a clean state");
    }

    // Figure out all of the coordinates we will be resolving
    let coordinates = perform("Enumerating coordinates", || {
        let mut coordinates = Vec::<String>::new();
        let mut merge_coordinates_from_layer = |layer: &Layer| {
            let coordinates_in_layer: Vec<String> = layer
                .coordinates()
                .iter()
                .map(|coord| coord.to_owned())
                .collect::<_>();
            coordinates.extend(coordinates_in_layer);
        };

        // Add mandatory layers
        let sets = LayerSets::new(&sparse_repo);
        let layer_set = sets
            .mandatory_layers()
            .context("resolving mandatory layers")?;
        for layer in layer_set.layers() {
            merge_coordinates_from_layer(layer);
        }

        if let Some(selected) = sets.selected_layers().context("loading selected layers")? {
            // Add selected layers' coordinates
            if selected.layers().is_empty() {
                eprintln!("No layers are selected, but a stack exists");
                bail!("No layers found");
            }
            for layer in selected.layers() {
                merge_coordinates_from_layer(layer);
            }
        } else {
            // Add ad-hoc layer coordinates
            if let Some(adhoc_layers) = sets.adhoc_layers().context("reading adhoc layers")? {
                for layer in adhoc_layers.layers() {
                    merge_coordinates_from_layer(layer);
                }
            } else {
                // Fail because there are no selected layers or ad-hoc layer
                eprintln!("There are no selected layers and an ad-hoc layer does not exist.");
                eprintln!("The focused development working state in this repo might be corrupted.");
                bail!("No layers found");
            }
        }
        Ok(coordinates)
    })?;

    let coordinate_set =
        CoordinateSet::try_from(coordinates.as_ref()).context("constructing coordinate set")?;

    let cloned_app = app.clone();
    let dense_revision = perform("Determining the current commit in the dense repo", || {
        git_helper::run_git_command_consuming_stdout(
            "Determining the current commit in the dense repo".to_owned(),
            &dense_repo,
            vec!["rev-parse", "HEAD"],
            cloned_app,
        )
    })?;

    let cloned_app = app.clone();
    let sparse_revision = perform("Determining the current commit in the sparse repo", || {
        git_helper::run_git_command_consuming_stdout(
            "Determining the current commit in the sparse repo".to_owned(),
            &sparse_repo,
            vec!["rev-parse", "HEAD"],
            cloned_app,
        )
    })?;

    perform("Backing up the current sparse checkout file", || {
        std::fs::copy(&sparse_profile_path, &sparse_checkout_backup_path)
            .context("copying to the backup file")?;

        Ok(())
    })?;

    let sparse_objects_directory = std::fs::canonicalize(&sparse_repo)
        .context("canonicalizing sparse path")?
        .join(".git")
        .join("objects");

    let cloned_app = app.clone();
    perform("Switching in the dense repo", || {
        // When checking out in the dense repo, we make the sparse repo objects available as an alternate so as to not need to push to the dense repo.
        sparse_repos::switch_to_detached_branch_discarding_changes(
            &dense_repo,
            &sparse_revision.as_str(),
            Some(sparse_objects_directory.as_ref()),
            cloned_app,
        )
    })?;

    let cloned_app = app.clone();
    perform("Computing the new sparse profile", || {
        sparse_repos::generate_sparse_profile(
            &dense_repo,
            &sparse_profile_output_path,
            coordinate_set,
            cloned_app,
        )
    })?;

    let cloned_app = app.clone();
    perform("Resetting in the dense repo", || {
        git_helper::run_git_command_consuming_stdout(
            "Resetting in the dense repo".to_owned(),
            &dense_repo,
            vec!["reset", "--hard", &dense_revision],
            cloned_app,
        )
    })?;

    let cloned_app = app.clone();
    if let Err(_e) = perform("Applying the sparse profile", || {
        sparse_repos::set_sparse_config(&sparse_repo, cloned_app.clone())?;

        let sparse_profile_output_file =
            File::open(&sparse_profile_output_path).context("opening new sparse profile")?;
        let (mut cmd, scmd) = SandboxCommand::new_with_handles(
            "Applying the sparse profile".to_owned(),
            git_helper::git_binary(),
            Some(Stdio::from(sparse_profile_output_file)),
            None,
            None,
            cloned_app.clone(),
        )?;
        scmd.ensure_success_or_log(
            cmd.current_dir(&sparse_repo)
                .arg("sparse-checkout")
                .arg("set")
                .arg("--stdin"),
            SandboxCommandOutput::Stderr,
            "sparse-checkout set",
        )
        .map(|_| ())
        .context("setting sparse checkout from new profile")
    }) {
        perform("Restoring and reapplying the backup profile", || {
            let backup_file = File::open(&sparse_checkout_backup_path)
                .context("opening backup sparse profile")?;
            let (mut cmd, scmd) = SandboxCommand::new_with_handles(
                "Restoring and reapplying the backup profile".to_owned(),
                git_helper::git_binary(),
                Some(Stdio::from(backup_file)),
                None,
                None,
                cloned_app.clone(),
            )?;
            scmd.ensure_success_or_log(
                cmd.current_dir(&sparse_repo)
                    .arg("sparse-checkout")
                    .arg("set")
                    .arg("--cone")
                    .arg("--stdin"),
                SandboxCommandOutput::Stderr,
                "sparse-checkout set",
            )
            .map(|_| ())
        })?;
    }

    let cloned_app = app.clone();
    perform("Updating the sync point", || {
        sparse_repos::configure_sparse_sync_point(&sparse_repo, cloned_app)
    })?;

    let cloned_app = app.clone();
    perform("Disabling the filesystem monitor", || {
        sparse_repos::config_sparse_disable_filesystem_monitor(&sparse_repo, cloned_app)
    })?;

    Ok(())
}
