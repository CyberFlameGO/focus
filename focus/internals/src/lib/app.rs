use std::borrow::Borrow;
use std::sync::Arc;

use crate::{ui::UserInterface, util::sandbox::Sandbox};
use anyhow::{Context, Result};
use std::time::SystemTime;
use ti_library::tool_insights_client::ToolInsightsClient;

#[derive(Clone)]
pub struct App {
    ui: Arc<UserInterface>,
    sandbox: Arc<Sandbox>,
    tool_insights_client: ToolInsightsClient,
}

impl App {
    pub fn new(preserve_sandbox_contents: bool, interactive: bool) -> Result<Self> {
        let ui = Arc::from(UserInterface::new(interactive).context("Failed to start UI")?);
        let sandbox =
            Arc::from(Sandbox::new(preserve_sandbox_contents).context("Failed to create sandbox")?);
        let tool_insights_client = ToolInsightsClient::new(
            // TODO: get this from toml file
            "focus".to_string(),
            // TODO: get this from toml file
            "0.3".to_string(),
            SystemTime::now(),
        );
        Ok(Self {
            ui,
            sandbox,
            tool_insights_client,
        })
    }

    /// Get a reference to the app's ui.
    pub fn ui(&self) -> Arc<UserInterface> {
        self.ui.clone()
    }

    /// Get a reference to the app's sandbox.
    pub fn sandbox(&self) -> Arc<Sandbox> {
        self.sandbox.clone()
    }

    /// Get a reference to the app's tool-insights client.
    pub fn tool_insights_client(&self) -> &ToolInsightsClient {
        self.tool_insights_client.borrow()
    }
}
