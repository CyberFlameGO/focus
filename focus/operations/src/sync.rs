// Copyright 2022 Twitter, Inc.
// SPDX-License-Identifier: Apache-2.0

use core::sync::atomic::AtomicBool;
use focus_internals::index::RocksDBMemoizationCacheExt;
use focus_internals::{locking, model::repo::Repo, target::TargetSet};

use crate::util::perform;
use content_addressed_cache::RocksDBCache;
use focus_util::app::App;
use focus_util::backed_up_file::BackedUpFile;
use tracing::{debug, info, warn};

use std::convert::TryFrom;

use std::path::Path;

use std::sync::Arc;
use std::time::{Duration, SystemTime};

use anyhow::{bail, Context, Result};
use lazy_static::lazy_static;

const PREEMPTIVE_SYNC_MAX_WAIT_MILLIS: u64 = 30000;
const TEST_ONLY_PREEMPTIVE_SYNC_MAX_WAIT_MILLIS_UNDER_TEST: u64 = 300;
const PREEMPTIVE_SYNC_POLL_INTERVAL_MILLIS: u64 = 100;

lazy_static! {
    static ref TEST_ONLY_PREEMPTIVE_SYNC_MACHINE_IS_ACTIVE: AtomicBool = AtomicBool::new(false);
}

pub fn test_only_get_preemptive_sync_machine_is_active() -> bool {
    use std::sync::atomic::Ordering;
    TEST_ONLY_PREEMPTIVE_SYNC_MACHINE_IS_ACTIVE.load(Ordering::SeqCst)
}

#[cfg(test)]
pub fn test_only_set_preemptive_sync_machine_is_active(new_value: bool) {
    use std::sync::atomic::Ordering;
    TEST_ONLY_PREEMPTIVE_SYNC_MACHINE_IS_ACTIVE.store(new_value, Ordering::SeqCst);
}

/// An enumeration capturing that the sync was peformed or a reason it was skipped.
#[derive(Debug, PartialEq, Eq)]
pub enum SyncStatus {
    /// The sync was performed.
    Success,

    /// There is no change to the sync point itself.
    SkippedSyncPointUnchanged,

    /// The content that has changed is not relevant to the build graph.
    SkippedSyncPointDifferenceIrrelevant,

    /// Preemptive syncing is not enabled.
    SkippedPreemptiveSyncDisabled,

    /// Preemptive syncing was cancelled because the machine is actively being used.
    SkippedPreemptiveSyncCancelledByActivity,
}

/// State describing the outcome of a sync.
pub struct SyncResult {
    /// Whether the working tree was checked out during the sync
    pub checked_out: bool,

    /// The commit ID that was synchronized
    pub commit_id: Option<git2::Oid>,

    /// The action taken
    pub status: SyncStatus,
}

/// Synchronize the sparse repo's contents with the build graph. Returns a SyncResult indicating what happened.
pub fn run(sparse_repo: &Path, preemptive: bool, app: Arc<App>) -> Result<SyncResult> {
    let repo = Repo::open(sparse_repo, app.clone()).context("Failed to open the repo")?;

    if preemptive {
        if !repo.get_preemptive_sync_enabled()? {
            return Ok(SyncResult {
                checked_out: false,
                commit_id: None,
                status: SyncStatus::SkippedPreemptiveSyncDisabled,
            });
        }

        let idle_duration = repo.get_preemptive_sync_idle_threshold()?;
        let max_wait = Duration::from_millis(if cfg!(test) {
            TEST_ONLY_PREEMPTIVE_SYNC_MAX_WAIT_MILLIS_UNDER_TEST
        } else {
            PREEMPTIVE_SYNC_MAX_WAIT_MILLIS
        });
        let poll_interval = Duration::from_millis(PREEMPTIVE_SYNC_POLL_INTERVAL_MILLIS);
        info!(
            ?idle_duration,
            ?max_wait,
            ?poll_interval,
            "Waiting for machine to be idle"
        );
        if wait_for_machine_to_be_idle(idle_duration, max_wait, poll_interval)
            .context("Failed waiting for machine to be idle")?
        {
            info!("Machine is idle, continuing preemptive sync");
        } else {
            info!("Machine is busy, cancelling preemptive sync");
            return Ok(SyncResult {
                checked_out: false,
                commit_id: None,
                status: SyncStatus::SkippedPreemptiveSyncCancelledByActivity,
            });
        }
    }

    let _lock = locking::hold_lock(sparse_repo, Path::new("sync.lock"))
        .context("Failed to obtain synchronization lock")?;

    let sparse_profile_path = repo.git_dir().join("info").join("sparse-checkout");
    if !sparse_profile_path.is_file() {
        bail!("This does not appear to be a focused repo -- it is missing a sparse checkout file");
    }

    let selections = repo.selection_manager()?;
    let selection = selections.computed_selection()?;
    let targets = TargetSet::try_from(&selection).context("constructing target set")?;

    // Add target/project to TI data.
    let app_for_ti_client = app.clone();
    let ti_client = app_for_ti_client.tool_insights_client();
    ti_client.get_context().add_to_custom_map(
        "sync_kind",
        if preemptive {
            "preemptive"
        } else {
            "immediate"
        },
    );

    let backed_up_sparse_profile: Option<BackedUpFile> = if preemptive {
        None
    } else {
        super::ensure_clean::run(sparse_repo, app.clone())
            .context("Failed trying to determine whether working trees were clean")?;

        ti_client
            .get_context()
            .add_to_custom_map("total_target_count", targets.len().to_string());
        ti_client.get_context().add_to_custom_map(
            "user_selected_project_count",
            selection.projects.len().to_string(),
        );
        ti_client.get_context().add_to_custom_map(
            "user_selected_target_count",
            selection.targets.len().to_string(),
        );

        Some(BackedUpFile::new(&sparse_profile_path)?)
    };

    let head_commit = repo.get_head_commit().context("Resolving head commit")?;

    // Figure out if this repo has a "master" branch or "main" branch.
    let primary_branch_name =
        primary_branch_name(&repo).context("Determining primary branch name")?;

    let commit = if preemptive {
        if let Some(prefetch_commit) = repo
            .get_prefetch_head_commit("origin", primary_branch_name.as_str())
            .context("Resolving prefetch head commit")?
        {
            prefetch_commit
        } else {
            bail!("No prefetch commit found for preemptive sync");
        }
    } else {
        head_commit
    };

    if preemptive {
        if let Some(working_tree) = repo.working_tree() {
            if let Ok(Some(sync_point)) = working_tree.read_sparse_sync_point_ref() {
                if sync_point == commit.id() {
                    // The sync point is already set to this ref. We don't need to bother.
                    warn!("Skipping preemptive synchronization because the commit to sync is the same as that of the sync point");
                    return Ok(SyncResult {
                        checked_out: false,
                        commit_id: Some(commit.id()),
                        status: SyncStatus::SkippedSyncPointUnchanged,
                    });
                }
            } else if let Ok(Some(sync_point)) = working_tree.read_preemptive_sync_point_ref() {
                if sync_point == commit.id() {
                    // The sync point is already set to this ref. We don't need to bother.
                    warn!("Skipping preemptive synchronization because the commit to sync is the same as that of the preemptive sync point");
                    return Ok(SyncResult {
                        checked_out: false,
                        commit_id: Some(commit.id()),
                        status: SyncStatus::SkippedSyncPointUnchanged,
                    });
                }
            }
        }
        // TODO: Skip outlining if there are no changes to the build graph between the last and new prospective sync point
    }

    let (pattern_count, checked_out) = perform("Computing the new sparse profile", || {
        let odb = RocksDBCache::new(repo.underlying());
        repo.sync(
            commit.id(),
            &targets,
            preemptive,
            &repo.config().index,
            app.clone(),
            &odb,
        )
        .context("Sync failed")
    })?;

    if preemptive {
        perform("Updating the sync point", || {
            repo.working_tree()
                .unwrap()
                .write_preemptive_sync_point_ref(commit.id())
        })?;
    } else {
        ti_client
            .get_context()
            .add_to_custom_map("pattern_count", pattern_count.to_string());
        perform("Updating the sync point", || {
            repo.working_tree().unwrap().write_sync_point_ref()
        })?;

        // The profile was successfully applied, so do not restore the backup.
        backed_up_sparse_profile.unwrap().set_restore(false);
    }

    Ok(SyncResult {
        checked_out,
        commit_id: Some(commit.id()),
        status: SyncStatus::Success,
    })
}

/// Wait for the machine to be idle for a given time period, waiting up to some maximum, and polling at a given interval.
fn wait_for_machine_to_be_idle(
    idle_duration: Duration,
    max_wait: Duration,
    poll_interval: Duration,
) -> Result<bool> {
    use focus_platform::session_state;

    if max_wait < idle_duration {
        bail!("max_wait must be greater than idle_duration")
    } else if poll_interval > max_wait {
        bail!("poll_interval must be less than max_wait")
    }

    let started_at = SystemTime::now();
    loop {
        let elapsed = started_at
            .elapsed()
            .context("Determining elapsed time failed")?;
        if elapsed > max_wait {
            break;
        }
        let state = {
            // If we are running under test, read from a variable instead of doing any polling.
            if cfg!(test) {
                debug!("Running under test!");
                if test_only_get_preemptive_sync_machine_is_active() {
                    debug!("Pretending machine is active");
                    session_state::SessionStatus::Active
                } else {
                    debug!("Pretending machine is idle");
                    session_state::SessionStatus::Idle
                }
            } else {
                unsafe { session_state::has_session_been_idle_for(idle_duration) }
            }
        };

        match state {
            session_state::SessionStatus::Active => {
                std::thread::sleep(poll_interval);
            }
            _ => {
                // Note: If we can't determine whether the session is idle, just go ahead.
                return Ok(true);
            }
        }
    }

    Ok(false)
}

fn primary_branch_name(repo: &Repo) -> Result<String> {
    let underlying = repo.underlying();
    if underlying.find_reference("refs/heads/master").is_ok() {
        Ok(String::from("master"))
    } else if underlying.find_reference("refs/heads/main").is_ok() {
        Ok(String::from("main"))
    } else {
        bail!("Could not determine primary branch name")
    }
}
