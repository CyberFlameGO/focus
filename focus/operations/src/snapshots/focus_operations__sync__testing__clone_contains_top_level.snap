---
source: focus/operations/src/sync.rs
expression: "&profile"
---
/*
!/*/
/focus/
*.bzl
BUILD*
WORKSPACE*

